### PR TITLE
[WIP] tf32 tensor core support [run_process_replay]

### PR DESCRIPTION
**Pull request by ignaciosica**
#### testing

I believe its reasonable to expect this error considering the use of tf32

```
Not equal to tolerance rtol=0.03, atol=0.0001

Mismatched elements: 72610 / 16777216 (0.433%)
Max absolute difference among violations: 50.879272
Max relative difference among violations: 0.05080053
 ACTUAL: array([[1010.2617 , 1022.77563, 1025.6896 , ..., 1015.4346 , 1020.4208 ,
        1030.051  ],
       [1004.36743, 1014.6489 , 1004.6634 , ...,  999.78436, 1017.1111 ,...
 DESIRED: array([[1000.8336 , 1011.0188 , 1010.40546, ..., 1008.2627 , 1015.9673 ,
        1027.1831 ],
       [1020.1254 , 1024.1477 , 1016.58563, ..., 1019.70245, 1020.1438 ,...
```

### CODE REVIEW

**Comment by geohot** [comment]
I think it should be gated behind TF32=1 for tests to pass

File: tinygrad/renderer/assembly.py [hunk 1]
```diff
@@ -58,12 +59,9 @@ def load_store_ptr_arithmetic(x:UOp, buf:UOp, alu:Optional[UOp]=None, const:Opti
 ])
 
 class PTXRenderer(Renderer):
-  device = "CUDA"
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:61]: Why was this removed?

**Comment by geohot** [comment]
Also, this needs to be tested in CI.

**Comment by geohot** [comment]
Err, I think this is wrong.
```
tiny@tiny18:~/tinygrad/examples/llm.c$ ./train_gpt2.py --num_iterations 10 --skip_test --sequence_length 1024
ram used:  0.55 GB, lm_head.weight                                    : 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 161/161 [00:00<00:00, 446.69it/s]
loaded weights in 360.95 ms, 0.55 GB loaded at 1.52 GB/s
loading cached tokens in /raid/downloads/761b5e0462a105a9cb10633be1997ea9
iteration 0, loss: 41.449692, time: 10063.220ms, 407 tok/s
iteration 1, loss: 41.385033, time: 4356.282ms, 940 tok/s
iteration 2, loss: 41.332657, time: 323.638ms, 12656 tok/s
iteration 3, loss: 41.289211, time: 276.582ms, 14809 tok/s
iteration 4, loss: 41.252991, time: 280.617ms, 14596 tok/s
iteration 5, loss: 41.258862, time: 275.887ms, 14846 tok/s
iteration 6, loss: 41.301083, time: 272.249ms, 15045 tok/s
iteration 7, loss: 41.347343, time: 271.825ms, 15068 tok/s
iteration 8, loss: 41.402363, time: 273.834ms, 14957 tok/s
iteration 9, loss: 41.461010, time: 280.151ms, 14620 tok/s
tiny@tiny18:~/tinygrad/examples/llm.c$ 
```

vs

```
tiny@tiny18:~/tinygrad/examples/llm.c$ ./train_gpt2.py --num_iterations 10 --skip_test --sequence_length 1024
ram used:  0.55 GB, lm_head.weight                                    : 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 161/161 [00:00<00:00, 780.96it/s]
loaded weights in 206.69 ms, 0.55 GB loaded at 2.65 GB/s
loading cached tokens in /raid/downloads/761b5e0462a105a9cb10633be1997ea9
iteration 0, loss: 4.579117, time: 4692.359ms, 872 tok/s
iteration 1, loss: 4.052018, time: 2505.980ms, 1634 tok/s
iteration 2, loss: 3.755559, time: 1302.798ms, 3144 tok/s
iteration 3, loss: 3.577645, time: 1257.918ms, 3256 tok/s
iteration 4, loss: 3.380209, time: 1261.750ms, 3246 tok/s
iteration 5, loss: 3.202756, time: 1257.087ms, 3258 tok/s
iteration 6, loss: 3.049728, time: 1250.883ms, 3274 tok/s
```

on non TF32 TC branch

File: tinygrad/renderer/assembly.py (hunk 1 above)
Comment (ignaciosica) [reply tinygrad/renderer/assembly.py:61]: I removed it as it was already as default on the constructor and nowhere in the codebase was accesses statically (`PTXRenderer.device`).

File: tinygrad/renderer/assembly.py [hunk 2]
```diff
@@ -60,10 +61,8 @@ def load_store_ptr_arithmetic(x:UOp, buf:UOp, alu:Optional[UOp]=None, const:Opti
 class PTXRenderer(Renderer):
   device = "CUDA"
   suffix = "PTX"
-  global_max = (2147483647, 65535, 65535)
-  local_max = (1024, 1024, 64)
-  shared_max = 49152
-  tensor_cores = [tc for tc in CUDARenderer.tensor_cores if tc.dtype_in == dtypes.half]
+  global_max, local_max, shared_max = CUDARenderer.global_max, CUDARenderer.local_max, CUDARenderer.shared_max
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:64]: Is this related to tf32? If not, it should be in its own PR

Comment (ignaciosica) [reply tinygrad/renderer/assembly.py:64]: No, I'll move it to another one

**Comment by ignaciosica** [comment]
I found the bug, I blindly trusted CI but I was wrong, the thread pattern is different for tf32. I made the pr a back draft until it's ready for review.

**Comment by ignaciosica** [comment]
Closing in favor of #8292
